### ISSUE TITLE
Support for relative home path in pyvenv.cfg

### ISSUE THREAD

**Issue by 67be7836-e06e-4862-b214-f432d535bdb5** [python/cpython#83650]
BPO | [39469](https://bugs.python.org/issue39469)
--- | :---
Nosy | @brettcannon, @vsajip, @eryksun, @zooba, @teruokun
PRs | <li>python/cpython#18213</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-01-27.20:58:08.673>
labels = ['interpreter-core', 'type-feature', '3.9']
title = 'Support for relative home path in pyvenv.cfg'
updated_at = <Date 2020-02-05.00:08:02.180>
user = 'https://github.com/teruokun'
```

bugs.python.org fields:
```python
activity = <Date 2020-02-05.00:08:02.180>
actor = 'steve.dower'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2020-01-27.20:58:08.673>
creator = 'Jeff.Edwards'
dependencies = []
files = []
hgrepos = []
issue_num = 39469
keywords = ['patch']
message_count = 6.0
messages = ['360800', '360886', '360887', '360893', '360908', '360909']
nosy_count = 5.0
nosy_names = ['brett.cannon', 'vinay.sajip', 'eryksun', 'steve.dower', 'Jeff.Edwards']
pr_nums = ['18213']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue39469'
versions = ['Python 3.9']
```

</p></details>

**Comment by ncoghlan** [comment]
While portable venvs aren't *general purpose* venvs, they do work if you purge the `bin`/`Scripts` folder of everything other than `python` itself, and only support running commands that are accessible via the `-m` switch.

For that case (where the directly executable files are deleted outright rather than attempting to fix them up), `pyvenv.cfg` is the only file that requires modification if the relocation process also changes the path to the base runtime installation.

**Comment by thesayyn** [comment]
FWIW, only allowing is a deal breaker for us when generating these venv using Bazel, as it prevents cache sharing of these files with other people. Relative paths are not new, and is well understood by most engineers already, and since venv is already an advanced concept, i don't see why this restriction can not be lifted.

**Comment by vsajip** [comment]
It's not clear what you mean when you say "only allowing is a deal breaker for us". Only allowing what? Sorry if I'm being dense.

**Comment by thesayyn** [comment]
> "only allowing is a deal breaker for us".
sorry a typo. i meant this is a deal breaker for us.

**Comment by Vigilans** [comment]
> > Suffice to say, is there a significant reason to not allow it?简而言之，是否有重要的理由不允许它？
> 
> It's poorly supported by packaging. In particular, relocating an environment isn't supported with entry-point scripts, which pip installs with a fully-qualified shebang. Moreover, entry-point scripts in Windows are created as exe files (e.g. "pip.exe") that embed the fully-qualified path of python[w].exe in the environment, plus a zipped __main__.py. For example, given an environment at "C:\Temp\env", running "C:\Temp\env\Scripts\pip.exe" in turn spawns a child process with the command line: "C:\Temp\env\Scripts\python.exe" "C:\Temp\env\Scripts\pip.exe". This breaks if the environment is renamed or relocated.

`uv venv` now supports `--relocatable` flag to not relying on any absolute path to virtual env to run entrypoint scripts: https://github.com/astral-sh/uv/pull/5515, but `python` executable's symlink and `pyvenv.cfg`'s `home`field is still absolute path to 
 base python's installation. 

There is no issue in changing `python` symlink to relative path, so the support of relative path for `home` field in `pyvenv.cfg` is the last piece to a portable virtual environment. 

For now, when `home` field is relative path, it is relative to the working directory of python process which easily varies. Making it relative to `pyvenv.cfg`'s location would not be a breaking change.

**Comment by thesayyn** [comment]
> For now, when home field is relative path, it is relative to the working directory of python process which easily varies.

I didn’t know this was even permitted. So relative path is permitted but resolved to the current pwd?

**Comment by Vigilans** [comment]
> > For now, when home field is relative path, it is relative to the working directory of python process which easily varies.
> 
> I didn’t know this was even permitted. So relative path is permitted but resolved to the current pwd?

Yes, and you will fail to start python if you are not running it in the same directory as `pyvenv.cfg`.

**Comment by ncoghlan** [comment]
> i don't see why this restriction can not be lifted.

The project where I needed this has been published as open source since I commented on this issue, so the relevant post-install script is now available for direct linking: https://github.com/lmstudio-ai/venvstacks/blob/036d4e2629424fcd24e11cdec7d02110e39f6666/src/venvstacks/_injected/postinstall.py#L55

The interesting part of that function for purposes of this discussion is the `deployed_path` nested function, as that very intentionally does NOT resolve symlinks when determining the absolute path to put in the venv config file. That behaviour is most definitely NOT how `venv` creation normally handles symlinks (because it's not how Python startup handles them):

```
$ cat example-venv/pyvenv.cfg
home = /home/acoghlan/devel
include-system-site-packages = false
version = 3.12.6
executable = /usr/bin/python3.12
command = /home/acoghlan/devel/python-symlink -m venv /home/acoghlan/devel/example-venv
```

So even if relative path support is added (with the meaning of "relative to the venv config file"), there's an open question around whether it should resolve symlinks or not when making the relative path absolute. It possibly *shouldn't* (for the same reasons `venvstacks` doesn't: some platform environments may break with dynamic lib loading issues if you do that), but that would be an oddity relative to the way interpreter startup normally handles symlinks.

That said, the issue is still open because we think it's a reasonable thing to add, it's just that none of the current maintainers have a sufficient motivation to work on it (my looking into it for `venvstacks` is likely the closest any of us have come in years, and even that potential use case turned out to be better served by an external post-install script).

**Comment by StanFromIreland** [comment]
I am closing this as a duplicate of #136051 as it up-to-date.
